code/game/machinery/vending/cm_vending.dm

```diff
@@ -902,108 +873,190 @@ IN_USE						used for vending/denying
+		var/p_amount = myprod[2]				//amount left
+		ui_categories += list(p_amount)
+	data["stock_listing"] = ui_categories
+	return data
 
+/obj/structure/machinery/cm_vending/sorted/tgui_interact(mob/user, datum/tgui/ui)
+	if(!ishuman(user))
+		return
+
+	ui = SStgui.try_update_ui(user, src, ui)
 	if (!ui)
-		ui = new(user, src, ui_key, "cm_vending_sorted.tmpl", name , 600, 700)
-		ui.set_initial_data(data)
+		ui = new(user, src, "VendingSorted", name)
 		ui.open()
-		ui.set_auto_update(0)		//here we need autoupdate because items can be vended by others and are limited
 
-/obj/structure/machinery/cm_vending/sorted/Topic(href, href_list)
+/obj/structure/machinery/cm_vending/sorted/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
 	. = ..()
+	if(.)
+		return
+	if(!(in_range(src, usr) && isturf(loc) && ishuman(usr)))
+		return
 
-	handle_topic(usr, href, href_list)
-
-/obj/structure/machinery/cm_vending/sorted/handle_topic(mob/user, href, href_list)
-	if(in_range(src, user) && isturf(loc) && ishuman(user))
-		user.set_interaction(src)
-		if(href_list["vend"])
-
+	usr.set_interaction(src)
+	var/mob/living/carbon/human/H = usr
+	switch (action)
+		if ("vend")
 			if(stat & IN_USE)
 				return
+			var/has_access = can_access(usr)
+			if (has_access[1] == FALSE)
+				to_chat(usr, SPAN_WARNING(has_access[2]))
+				vend_fail()
+				return
 
-			var/mob/living/carbon/human/H = user
-
-			if(!hacked)
-
-				if(!allowed(H))
-					to_chat(H, SPAN_WARNING("Access denied."))
-					vend_fail()
-					return
-
-				var/obj/item/card/id/I = H.wear_id
-				if(!istype(I))
-					to_chat(H, SPAN_WARNING("Access denied. No ID card detected"))
-					vend_fail()
-					return
-
-				if(I.registered_name != H.real_name)
-					to_chat(H, SPAN_WARNING("Wrong ID card owner detected."))
-					vend_fail()
-					return
-
-				if(LAZYLEN(vendor_role) && !vendor_role.Find(H.job))
-					to_chat(H, SPAN_WARNING("This machine isn't for you."))
-					vend_fail()
-					return
-
-			var/idx=text2num(href_list["vend"])
-			var/list/topic_listed_products = get_listed_products(user)
+			var/idx=params["prod_index"]
+			var/list/topic_listed_products = get_listed_products(usr)
 			var/list/L = topic_listed_products[idx]
 
 			var/turf/T = get_appropriate_vend_turf(H)
 			if(T.contents.len > 25)
-				to_chat(H, SPAN_WARNING("The floor is too cluttered, make some space."))
+				to_chat(usr, SPAN_WARNING("The floor is too cluttered, make some space."))
 				vend_fail()
 				return
 
 			if(L[2] <= 0)	//to avoid dropping more than one product when there's
-				to_chat(H, SPAN_WARNING("[L[1]] is out of stock."))
+				to_chat(usr, SPAN_WARNING("[L[1]] is out of stock."))
 				vend_fail()
 				return		// one left and the player spam click during a lagspike.
 
 			vend_succesfully(L, H, T)
 
-		add_fingerprint(user)
-		ui_interact(user) //updates the nanoUI window
+	add_fingerprint(usr)
+	tgui_interact(usr, ui, state)
```

Comment:
```suggestion
```
just make all of the returns above return TRUE, this triggers an update

Reply:
I've gone and made these changes